File: Chapter 02/FactoryMethodDesignPatternExample.java
interface ProductExample {
    void produce();
}


class ConcreteProductOne implements ProductExample {
    public void produce() {
        System.out.println("Producing Concrete Product One.");
    }
}
class ConcreteProductTwo implements ProductExample {
    public void produce() {
        System.out.println("Producing Concrete Product Two.");
    }
}


abstract class CreatorExample {
    abstract ProductExample factoryMethod();

    void operation() {
        ProductExample product = factoryMethod();
        product.produce();
    }
}

class ConcreteCreatorOne extends CreatorExample {
    Product factoryMethod() {
        return new ConcreteProductOne();
    }
}

class ConcreteCreatorTwo extends CreatorExample {
    Product factoryMethod() {
        return new ConcreteProductTwo();
    }
}


public class FactoryMethodDesignPatternExample {
    public static void main(String[] args) {
        CreatorExample creatorOne = new ConcreteCreatorOne();
        CreatorExample creatorTwo = new ConcreteCreatorTwo();

        creatorOne.operation();
        creatorTwo.operation();
    }
}

File: Chapter 02/FactoryMethodDesignPatternExample.py
class ProductExample:
    def produce(self):
        pass
class ConcreteProductOne(ProductExample):
    def produce(self):
        print("Producing Concrete Product One.")
class ConcreteProductTwo(ProductExample):
    def produce(self):
        print("Producing Concrete Product Two.")
class CreatorExample:
    def factory_method(self):
        pass
    def operation(self):
        product = self.factory_method()
        product.produce()
class ConcreteCreatorOne(CreatorExample):
    def factory_method(self):
        return ConcreteProductOne()

class ConcreteCreatorTwo(CreatorExample):
    def factory_method(self):
        return ConcreteProductTwo()

if __name__ == "__main__":
    creatorOne = ConcreteCreatorOne()
    creatorTwo = ConcreteCreatorTwo()

    creatorOne.operation()
    creatorTwo.operation()
